LabsPlus/4you-api
# Feature/add customer validation

Description:

Added customer validations using Yup for create and update actions.
Integrated bcrypt for password hashing during customer creation and validation during login.
Implemented JWT token generation for user authentication.
Updated the database schema to use Date for birth_date and payment_date fields.
Changed the update method for customer data from PUT to PATCH for partial updates.

Changes:

CustomerController:

Added validation logic using Yup for customer creation and update.
Integrated bcrypt for password encryption and comparison.
Added JWT generation for login.
Changed the update route to PATCH for updating customer data partially.

Database Schema:

Updated the Customer and Payment models to use Date type for birth_date and payment_date fields.

Routes:

Changed the PUT method to PATCH for updatingCustomer data in customer.routes.js.
Packages:

Added bcryptjs for password hashing.
Added jose for JWT token handling.
Added yup for validation.

How to Test:

Customer Creation:

Send a POST request to /customer/create with the following body:
json

{
  "name": "John Doe",
  "email": "john.doe@example.com",
  "password": "StrongPass1!",
  "phone": "+5534567890",
  "taxid": "123.456.789-00",
  "birth_date": "1990-01-01"
}
The password should be hashed before storing in the database.
Ensure the response contains the newly created customer data.
Customer Update:

Send a PATCH request to /customer/{email} with the updated customer data (e.g., phone number or email).
json

{
  "phone": "+0987654321"
}
Ensure that the customer data is updated correctly and returned in the response.
Customer Login:

Send a POST request to /customer/login with the following body:
json

{
  "email": "john.doe@example.com",
  "password": "StrongPass1!"
}
Ensure that the response contains a valid JWT token if the credentials are correct.
If the credentials are incorrect, ensure the response returns an error with status 401.
Database Changes:

Verify that the birth_date and payment_date fields are stored as Date types in the database.
